src/core/vector/qgsvectorlayereditutils.cpp

```diff
@@ -735,6 +735,119 @@ int QgsVectorLayerEditUtils::addTopologicalPoints( const QgsPointXY &p )
   return addTopologicalPoints( QgsPoint( p ) );
 }
 
+bool QgsVectorLayerEditUtils::mergeFeatures( const QgsFeatureIds &mergeFeatureIds, const QgsAttributes &mergeAttributes, QgsGeometry unionGeometry, QString &errorMessage )
```

Comment:
Could we alterantively do the following:

- have explicitly a "target feature" being passed into this method
- and 1 or more "merge features"
- we *always* merge into and update that target feature

This basically would get rid of the heuristics in here to determine the target feature by id and some codepaths (when to update or insert)